### ISSUE TITLE
LIKE matches till new line

### ISSUE THREAD

**Issue by kgpai** [prestodb/presto#23281]
## Your Environment

* Presto version used: Mainline Presto

## Current Behavior

It seems when using like Presto only matches till the new line. 

-- new line is important below

```
presto:di> select case when x like 'test % test' then 'match' else 'no match' end from (values ('test abc test
           ->   xyz')) t(x);
 _col0                
-------
 match 
(1 row)

```

 This results in inconsistencies like so

```
presto:di> select 'abc' || chr(10) || '123' like 'abc';
 _col0
-------
 true

presto:di> select 'abc' || chr(10) || '123' = 'abc';
 _col0
-------
 false

```
## Expected Behavior
Presto should take the whole string into consideration and not just till the new line. 

## Context
This was discovered during Presto and Prestissimo verification.

**Comment by kgpai** [comment]
cc: @mbasmanova @kaikalur @kagamiori @tdcmeehan @amitkdutta

**Comment by kgpai** [comment]
cc: @rschlussel

**Comment by mbasmanova** [comment]
CC: @elharo

**Comment by kaikalur** [comment]
Looks like constant folding issue. Doing it in a lambda works fine:

```
presto:di> select transform(array['abc' || chr(10) || '123'], x->x like 'abc');
  _col0
---------
 [false]
(1 row)
```

**Comment by kaikalur** [comment]
Even simpler:

```
presto:di> select x like 'abc' from (select 'abc' || chr(10) || '123' x);
 _col0
-------
 false
(1 row)
```

**Comment by tdcmeehan** [comment]
@kaikalur I don't think this is a problem specific to constant folding.  Load up the memory connector and execute the following:

`create table memory.default.test (val varchar);`

```
insert into memory.default.test (val) VALUES ('abc
     -> 123');
```

`select val LIKE 'a%c' from memory.default.test;`

This erroneously returns true.  I think it's a problem with the like operator, likely a bug in the regex library or how we're using it.

**Comment by kaikalur** [comment]
Hmm ok - wonder if we need to set multiline explicitly for joni regexp when we compile LIKE to regexp

**Comment by kaikalur** [comment]
It gets weirder:

```
presto:di> select x like 'a%c' from (select 'abcd' x union all select  'abcd'||chr(10)||'xyz');
 _col0
-------
 false
 false
(2 rows)
```

**Comment by kaikalur** [comment]
it works fine here even with a newline. I wonder if connector pushdown is an issue (though I would think there is no special connector for memory)

**Comment by kaikalur** [comment]
So trying to define the issue more cleanly - a) if the last char is not % and b) if the last char is the end of the (first) line, it will match upto that only and stop

**Comment by kgpai** [comment]
@kaikalur I have seen a few cases where queries inadvertantly rely on this behavior , not sure if its intentional or unintentional.

**Comment by amitkdutta** [comment]
> @kaikalur I have seen a few cases where queries inadvertantly rely on this behavior , not sure if its intentional or unintentional.

I think people rely on Presto and whatever it gives they think it as correct. So they build parsing around it, but for legitimate issues we need to fix the bug (like this one) and make customers aware that the change is coming.

**Comment by jp-sivaprasad** [comment]
Fixed by [#23404](https://github.com/prestodb/presto/pull/23404)
```
Suggested tests:
@Test
public void testLikeNoMatchWithNewLine()
{
    Regex regex = likePattern(utf8Slice("foo"));
    assertFalse(likeVarchar(utf8Slice("foo\nbar"), regex));
}

@Test
public void testLikeNewlineAtMatch()
{
    Regex regex = likePattern(utf8Slice("foo_bar"));
    assertTrue(likeVarchar(utf8Slice("foo\nbar"), regex));
}

@Test
public void testLikeNewlineInMatchWithSingle()
{
    Regex regex = likePattern(utf8Slice("fo_%"));
    assertTrue(likeVarchar(utf8Slice("foo\nbar"), regex));
}
```
